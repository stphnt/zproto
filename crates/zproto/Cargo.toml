[package]
name = "zproto"
version = "0.4.0-pre"
authors = ["Stephen Hunt <stphnhnt.git@gmail.com>"]
edition = "2021"
description = "A library from communicating with Zaber products in Rust."
license = "MIT"
readme = "README.md"
repository = "https://github.com/stphnt/zproto"
documentation = "https://docs.rs/zproto"
homepage = "https://github.com/stphnt/zproto"
keywords = ["Zaber", "ASCII", "Binary", "serial", "RS232"]
categories = ["hardware-support"]
include = [
    "examples/",
    "src/",
    "test/",
    "LICENSE",
    "README.md",
]

[package.metadata.docs.rs]
all-features = true

[features]
default = ["ascii", "binary", "v_latest"]
ascii = [] # Include the ASCII protocol
binary = [] # Include the Binary protocol
v6_14 = [] # Enable APIs related to firmware version 6.14
v6_15 = [] # Enable APIs related to firmware version 6.15
v6_16 = [] # Enable APIs related to firmware version 6.16
v6_17 = [] # Enable APIs related to firmware version 6.17
v6_18 = [] # Enable APIs related to firmware version 6.18
v6_19 = [] # Enable APIs related to firmware version 6.19
v6_20 = [] # Enable APIs related to firmware version 6.20
v6_21 = [] # Enable APIs related to firmware version 6.21
v6_22 = [] # Enable APIs related to firmware version 6.22
v6_23 = [] # Enable APIs related to firmware version 6.23
v6_24 = [] # Enable APIs related to firmware version 6.24
v6_25 = [] # Enable APIs related to firmware version 6.25
v6_26 = [] # Enable APIs related to firmware version 6.26
v6_27 = [] # Enable APIs related to firmware version 6.27
v6_28 = [] # Enable APIs related to firmware version 6.28
v6_29 = [] # Enable APIs related to firmware version 6.29
v6_30 = [] # Enable APIs related to firmware version 6.30
v6_31 = [] # Enable APIs related to firmware version 6.31
v6_32 = [] # Enable APIs related to firmware version 6.32
v7_01 = [] # Enable APIs related to firmware version 7.01
v7_03 = [] # Enable APIs related to firmware version 7.03
v7_05 = [] # Enable APIs related to firmware version 7.05
v7_06 = [] # Enable APIs related to firmware version 7.06
v7_07 = [] # Enable APIs related to firmware version 7.07
v7_08 = [] # Enable APIs related to firmware version 7.08
v7_09 = [] # Enable APIs related to firmware version 7.09
v7_10 = [] # Enable APIs related to firmware version 7.10
v7_11 = [] # Enable APIs related to firmware version 7.11
v7_12 = [] # Enable APIs related to firmware version 7.12
v7_13 = [] # Enable APIs related to firmware version 7.13
v7_14 = [] # Enable APIs related to firmware version 7.14
v7_15 = [] # Enable APIs related to firmware version 7.15
v7_16 = [] # Enable APIs related to firmware version 7.16
v7_17 = [] # Enable APIs related to firmware version 7.17
v7_18 = [] # Enable APIs related to firmware version 7.18
v7_19 = [] # Enable APIs related to firmware version 7.19
v7_20 = [] # Enable APIs related to firmware version 7.20
v7_21 = [] # Enable APIs related to firmware version 7.21
v7_22 = [] # Enable APIs related to firmware version 7.22
v7_23 = [] # Enable APIs related to firmware version 7.23
v7_24 = [] # Enable APIs related to firmware version 7.24
v7_25 = [] # Enable APIs related to firmware version 7.25
v7_26 = [] # Enable APIs related to firmware version 7.26
v7_27 = [] # Enable APIs related to firmware version 7.27
v7_28 = [] # Enable APIs related to firmware version 7.28
v7_29 = [] # Enable APIs related to firmware version 7.29
v7_30 = [] # Enable APIs related to firmware version 7.30
v7_31 = [] # Enable APIs related to firmware version 7.31
v7_32 = [] # Enable APIs related to firmware version 7.32
v7_33 = [] # Enable APIs related to firmware version 7.33
v7_34 = [] # Enable APIs related to firmware version 7.34
v7_35 = [] # Enable APIs related to firmware version 7.35
v7_36 = [] # Enable APIs related to firmware version 7.36
v7_37 = [] # Enable APIs related to firmware version 7.37
v7_38 = [] # Enable APIs related to firmware version 7.38
v7_39 = [] # Enable APIs related to firmware version 7.39
v_latest = [] # Enable APIs related to the latest firmware version

# Use the nightly-only `doc_cfg` feature when generating the documentation.
doc_cfg = []

[dependencies]
log = "0.4.19"
paste = "1.0.12"
serialport = "4.2.1"

[dev-dependencies]
simple_logger = "4.2.0"
static_assertions = "1.1.0"
trybuild = "1.0.80"

[[example]]
name = "ascii"
required-features = ["ascii"]

[[example]]
name = "binary"
required-features = ["binary"]
