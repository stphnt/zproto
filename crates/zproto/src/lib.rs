//! A simple library for communicating with Zaber products.
//!
//! Both Zaber's ASCII and Binary protocols are supported over serial and TCP
//! ports. See the [`ascii`] and [`binary`] modules for an introduction to their
//! use.

#![cfg_attr(all(doc, feature = "doc_cfg"), feature(doc_cfg))]

#[cfg(feature = "ascii")]
#[cfg_attr(all(doc, feature = "doc_cfg"), doc(cfg(feature = "ascii")))]
pub mod ascii;
pub mod backend;
#[cfg(feature = "binary")]
#[cfg_attr(all(doc, feature = "doc_cfg"), doc(cfg(feature = "binary")))]
pub mod binary;
pub mod convert;
pub mod error;
pub mod routine;
pub mod timeout_guard;

#[cfg(not(any(feature = "ascii", feature = "binary")))]
compile_error!("At least one of the `ascii` or `binary` features must be specified");

// Check the contents of the README (but don't include it in the docs).
#[doc = include_str!("../README.md")]
#[allow(unused_doc_comments)]
extern "C" {}
