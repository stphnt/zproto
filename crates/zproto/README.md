# zproto

A Rust implementation of Zaber's ASCII and Binary Protocols.

This library is unofficial. [Zaber Motion Library](https://www.zaber.com/software)
is Zaber's official communications library, but as of this writing there are no Rust bindings.

## Usage

Add this to your `Cargo.toml:

```toml
[dependencies]
zproto = "0.4.0-pre"
```

## Getting started

This library aims to be simple but robust. Communicating with a product in Zaber's
ASCII protocol looks something like this:

```rust,no_run
use zproto::{
    ascii::{response::check, Port},
    error::Error,
};

fn main() -> Result<(), Error> {
    // Open the port, home device 1, and wait for it to finish.
    let mut port = Port::open_serial("/dev/ttyUSB0")?;
    port.command_reply((1, "home"))?.flag_ok_and(
        // Ignore warnings about it being unhomed.
        check::warning_in(("WR", "WH", "--"))
    )?;
    port.poll_until_idle(1, check::flag_ok())?;

    // Move towards the end of travel and monitor position as it goes.
    // Once the position exceeds 100000, interrupt the motion.
    port.command_reply((1, "move max"))?.flag_ok()?;
    port.poll_until((1, "get pos"), check::flag_ok(), |reply| {
        let pos: i32 = reply.data().parse().unwrap();
        pos >= 100_000
    })?;
    port.command_reply((1, "stop"))?.flag_ok_and(check::warning_is("NI"))?;
    Ok(())
}
```

See the [`ascii`](https://docs.rs/zproto/latest/zproto/ascii) or
[`binary`](https://docs.rs/zproto/latest/zproto/binary) module documentation for
a more in-depth introduction to communicating with devices using the Zaber ASCII
or Binary protocols.

The [examples](examples) folder has some simple applications.

## Documentation

The documentation is available on [docs.rs](https://docs.rs/zproto).

## Cargo Features

By default, both the ASCII and Binary protocols are enabled via the `ascii` and
`binary` Cargo features, respectively. However, if you only want to use one,
`ascii` for example, you can specify that in your `Cargo.toml`:

```toml
[dependencies]
zproto = { version = "0.4.0-pre", default-features = false, features = ["ascii"] }
```

This will only include portions of the library related to the ASCII protocol
during compilation.

## License

This project is license under the [MIT License](LICENSE).

## Troubleshooting

1. Building fails with an error about `libudev`

    This crate requires the `libudev` shared library to be installed on your system, which some operating systems do not have installed by default.
    Installing the library will solve the problem.
    For example, on Ubuntu the library can be installed with `sudo apt install libudev-dev`.
