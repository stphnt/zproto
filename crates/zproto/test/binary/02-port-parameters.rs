//! Tests to confirm that port methods accept/reject commands appropriately at
//! compile time.
use zproto::{
    backend::Backend,
    binary::{command::*, Port}
};

fn _fail<B: Backend>(port: &mut Port<B>) -> Result<(), Box<dyn std::error::Error>> {
    let _ = port.tx_recv((0, RESET))?;  // Shouldn't work with tx_recv
    let _ = port.tx_recv((0, HOME, 0i32))?;  // Shouldn't take parameters
    let _ = port.tx_recv((0, MOVE_ABSOLUTE))?;  // Should take parameters
    let _ = port.tx_recv((0, MOVE_ABSOLUTE, false))?;  // Wrong parameter type

    let _ = port.tx_recv((0, untyped::HOME))?;  // u8 commands require an explicit data value

    port.tx((0, ERROR))?;  // Reply-only commands cannot be transmitted
    port.tx((0, ERROR, 0i32))?;  // Reply-only commands cannot be transmitted
    Ok(())
}

fn _ok<B: Backend>(port: &mut Port<B>) -> Result<(), Box<dyn std::error::Error>> {
    port.tx((1, RESET))?;  // RESET can be transmitted
    let _ = port.recv(MANUAL_MOVE_TRACKING)?; // Reply-only commands can be received
    Ok(())
}

fn main() {}
