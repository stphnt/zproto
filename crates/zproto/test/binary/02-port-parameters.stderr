error[E0277]: the trait bound `(u8, Reset): ElicitsResponse` is not satisfied
 --> test/binary/02-port-parameters.rs:9:26
  |
9 |     let _ = port.tx_recv((0, RESET))?;  // Shouldn't work with tx_recv
  |                  ------- ^^^^^^^^^^ the trait `ElicitsResponse` is not implemented for `(u8, Reset)`
  |                  |
  |                  required by a bound introduced by this call
  |
  = help: the following other types implement trait `ElicitsResponse`:
            (u8, Activate)
            (u8, Activate, D)
            (u8, ConvertToAscii)
            (u8, ConvertToAscii, D)
            (u8, EchoData)
            (u8, EchoData, D)
            (u8, Home)
            (u8, Home, D)
          and $N others
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx_recv`
 --> src/binary/port.rs
  |
  |     pub fn tx_recv<M>(&mut self, message: M) -> Result<Message<M::Response>, BinaryError>
  |            ------- required by a bound in this associated function
  |     where
  |         M: traits::TxMessage + traits::ElicitsResponse,
  |                                ^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx_recv`

error[E0277]: the trait bound `Home: TakesData<i32>` is not satisfied
  --> test/binary/02-port-parameters.rs:10:30
   |
10 |     let _ = port.tx_recv((0, HOME, 0i32))?;  // Shouldn't take parameters
   |                  -------     ^^^^ the trait `TakesData<i32>` is not implemented for `Home`, which is required by `(u8, Home, i32): TxMessage`
   |                  |
   |                  required by a bound introduced by this call
   |
   = help: the following other types implement trait `TakesData<T>`:
             <ConvertToAscii as TakesData<i32>>
             <EchoData as TakesData<i32>>
             <MoveAbsolute as TakesData<i32>>
             <MoveAtConstantSpeed as TakesData<i32>>
             <MoveIndex as TakesData<i32>>
             <MoveRelative as TakesData<i32>>
             <MoveToStoredPosition as TakesData<i32>>
             <ReadAnalogInput as TakesData<i32>>
           and $N others
   = note: required for `(u8, Home, i32)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx_recv`
  --> src/binary/port.rs
   |
   |     pub fn tx_recv<M>(&mut self, message: M) -> Result<Message<M::Response>, BinaryError>
   |            ------- required by a bound in this associated function
   |     where
   |         M: traits::TxMessage + traits::ElicitsResponse,
   |            ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx_recv`

error[E0277]: the trait bound `MoveAbsolute: TakesNoData` is not satisfied
  --> test/binary/02-port-parameters.rs:11:30
   |
11 |     let _ = port.tx_recv((0, MOVE_ABSOLUTE))?;  // Should take parameters
   |                  -------     ^^^^^^^^^^^^^ the trait `TakesNoData` is not implemented for `MoveAbsolute`, which is required by `(u8, MoveAbsolute): TxMessage`
   |                  |
   |                  required by a bound introduced by this call
   |
   = help: the following other types implement trait `TakesNoData`:
             Activate
             Home
             ReadAllDigitalInputs
             ReadAllDigitalOutputs
             Reset
             ReturnAnalogInputCount
             ReturnCalibratedEncoderCount
             ReturnCalibrationError
           and $N others
   = note: required for `(u8, MoveAbsolute)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx_recv`
  --> src/binary/port.rs
   |
   |     pub fn tx_recv<M>(&mut self, message: M) -> Result<Message<M::Response>, BinaryError>
   |            ------- required by a bound in this associated function
   |     where
   |         M: traits::TxMessage + traits::ElicitsResponse,
   |            ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx_recv`

error[E0277]: the trait bound `MoveAbsolute: TakesData<bool>` is not satisfied
  --> test/binary/02-port-parameters.rs:12:30
   |
12 |     let _ = port.tx_recv((0, MOVE_ABSOLUTE, false))?;  // Wrong parameter type
   |                  -------     ^^^^^^^^^^^^^ the trait `TakesData<bool>` is not implemented for `MoveAbsolute`, which is required by `(u8, MoveAbsolute, bool): TxMessage`
   |                  |
   |                  required by a bound introduced by this call
   |
   = help: the trait `TakesData<i32>` is implemented for `MoveAbsolute`
   = help: for that trait implementation, expected `i32`, found `bool`
   = note: required for `(u8, MoveAbsolute, bool)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx_recv`
  --> src/binary/port.rs
   |
   |     pub fn tx_recv<M>(&mut self, message: M) -> Result<Message<M::Response>, BinaryError>
   |            ------- required by a bound in this associated function
   |     where
   |         M: traits::TxMessage + traits::ElicitsResponse,
   |            ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx_recv`

error[E0277]: the trait bound `u8: TakesNoData` is not satisfied
  --> test/binary/02-port-parameters.rs:14:30
   |
14 |     let _ = port.tx_recv((0, untyped::HOME))?;  // u8 commands require an explicit data value
   |                  -------     ^^^^^^^^^^^^^ the trait `TakesNoData` is not implemented for `u8`, which is required by `(u8, u8): TxMessage`
   |                  |
   |                  required by a bound introduced by this call
   |
   = help: the following other types implement trait `TakesNoData`:
             Activate
             Home
             ReadAllDigitalInputs
             ReadAllDigitalOutputs
             Reset
             ReturnAnalogInputCount
             ReturnCalibratedEncoderCount
             ReturnCalibrationError
           and $N others
   = note: required for `(u8, u8)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx_recv`
  --> src/binary/port.rs
   |
   |     pub fn tx_recv<M>(&mut self, message: M) -> Result<Message<M::Response>, BinaryError>
   |            ------- required by a bound in this associated function
   |     where
   |         M: traits::TxMessage + traits::ElicitsResponse,
   |            ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx_recv`

error[E0277]: the trait bound `zproto::binary::command::types::Error: TakesNoData` is not satisfied
  --> test/binary/02-port-parameters.rs:16:17
   |
16 |     port.tx((0, ERROR))?;  // Reply-only commands cannot be transmitted
   |          --     ^^^^^ the trait `TakesNoData` is not implemented for `zproto::binary::command::types::Error`, which is required by `({integer}, zproto::binary::command::types::Error): TxMessage`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the following other types implement trait `TakesNoData`:
             Activate
             Home
             ReadAllDigitalInputs
             ReadAllDigitalOutputs
             Reset
             ReturnAnalogInputCount
             ReturnCalibratedEncoderCount
             ReturnCalibrationError
           and $N others
   = note: required for `(u8, zproto::binary::command::types::Error)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx`
  --> src/binary/port.rs
   |
   |     pub fn tx<M: traits::TxMessage>(&mut self, message: M) -> Result<Option<u8>, BinaryError> {
   |                  ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx`

error[E0277]: the trait bound `zproto::binary::command::types::Error: TakesData<i32>` is not satisfied
  --> test/binary/02-port-parameters.rs:17:17
   |
17 |     port.tx((0, ERROR, 0i32))?;  // Reply-only commands cannot be transmitted
   |          --     ^^^^^ the trait `TakesData<i32>` is not implemented for `zproto::binary::command::types::Error`, which is required by `({integer}, zproto::binary::command::types::Error, i32): TxMessage`
   |          |
   |          required by a bound introduced by this call
   |
   = help: the following other types implement trait `TakesData<T>`:
             <ConvertToAscii as TakesData<i32>>
             <EchoData as TakesData<i32>>
             <MoveAbsolute as TakesData<i32>>
             <MoveAtConstantSpeed as TakesData<i32>>
             <MoveIndex as TakesData<i32>>
             <MoveRelative as TakesData<i32>>
             <MoveToStoredPosition as TakesData<i32>>
             <ReadAnalogInput as TakesData<i32>>
           and $N others
   = note: required for `(u8, zproto::binary::command::types::Error, i32)` to implement `TxMessage`
note: required by a bound in `zproto::binary::Port::<'a, B, H>::tx`
  --> src/binary/port.rs
   |
   |     pub fn tx<M: traits::TxMessage>(&mut self, message: M) -> Result<Option<u8>, BinaryError> {
   |                  ^^^^^^^^^^^^^^^^^ required by this bound in `Port::<'a, B, H>::tx`
